File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();

        for (String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars); // <- TC : O( k log k)

            String key = new String(chars);
            // Make sure there is arraylist for every keyword to add word
            groups.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(groups.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
      # Creating a dict with default value as List
        groups = defaultdict(list) 

        for word in strs:
          # sorting the chars in str and then convert char list into str
          keyword = "".join(sorted(word))
          groups[keyword].append(word)
        
        return list(groups.values())